### ISSUE TITLE
XLA: Won't converge and doesn't respect visible devices.

### ISSUE THREAD

**Issue by joeyearsley** [tensorflow/tensorflow#16045]
------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes (writing a copy I can share)
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Nvidia Optimized Container 17.12
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: 1.4.1
- **Python version**: IntelPython 3.5
- **Bazel version (if compiling from source)**: 0.5.4
- **GCC/Compiler version (if compiling from source)**: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
- **CUDA/cuDNN version**: 9.0.176 with CUBLAS Basic Accelerated Linear Algebra 9.0.234 / 7
- **GPU model and memory**: Pascal Titan 12Gb
- **Exact command to reproduce**: Coming

### Describe the problem
I've seen 2 problems with XLA:
1) XLA doesn't respect visible devices (neither CUDA Visible devices or tf config options)
2) XLA doesn't converge

_1_
XLA doesn't respect visible devices, I have seen this in both Horovod and when running on a single GPU (on a Multi-GPU desktop), `N` XLA instances are created despite only one GPU being used.

_2_ 
When I run my script with and without XLA, the XLA compiled version plateaus and doesn't converge, whilst the none-XLA version converges. Where XLA has a loss of a magnitude higher.
The benchmark is a VGG network (with BatchNorm) and achieves a 3x speed up with XLA, it just doesn't converge.

Is this a known issue? Or is it just a possibility of occurring?

**Comment by shivaniag** [comment]
@jlebar do you know about these problems with XLA?

**Comment by jlebar** [comment]
(For future reference, it's a lot better to file two separate issues if you have two problems.  Otherwise when the bug becomes long, it's hard to track the problems.)

I use CUDA_VISIBLE_DEVICES with XLA all the time and it works fine.  I'm pretty surprised to hear that it's not working for you, because it takes effect in an nvidia library that's much lower-level than TF or XLA.  That is, because neither TF nor XLA is responsible for respecting this env var, I suspect the environment variable is getting stripped out somewhere in your setup.  Without steps to reproduce, I can't say more.

I'm not a TF person, so I can't speak to the TF config options not working, but again, I think a TF person would need additional information, probably concrete steps to reproduce, in order to assist.

With respect to your model not converging, that's very concerning.  But again I'd need concrete steps to reproduce in order to debug this.

**Comment by joeyearsley** [comment]
Having debugged _1_ a bit more, it seems that Tensorflow (or underlying libraries) does not totally respect the `config.gpu_options.visible_device_list`.
This results in a minor `10Mb` being taken when not using XLA and `500Mb` when using XLA on GPUs which are not in use in that process. I.e. Having an 8 Card system and wanting to run 4 jobs results in memory issues.

**Comment by joeyearsley** [comment]
Thank you, I'll try running with those suggestions!

**Comment by joeyearsley** [comment]
Did you mean CUDA 9.0.176? - In which case we have that version and still get the result.

I'll try CUDA 8 + cudnn 6, but I'll have to run that on a desktop as it doesn't support the new voltas (the build fails due to compute capability).

**Comment by jlebar** [comment]
9.0.176 is known bad, confirmed on our end.  nvidia tells us that 9.0.276 is the first 9.0.x version with the fix, but that the first public version with the fix is 9.0.282.

**Comment by joeyearsley** [comment]
Thanks! I’ll chase them up as I can’t find it on their site.

However, I also confirmed I still have an issue on TF 1.5rc and Cuda/Cudnn 8/6. 

Would a debug report like #11564 help? - I’ll hopefully have a script ready next week to recreate the issue.

**Comment by jlebar** [comment]
> Would a debug report like #11564 help?

Yes, it would.

**Comment by joeyearsley** [comment]
Here are the logs with the following flags: with "--tf_xla_parallel_checking=true --parallel_check_failfast=false"

[vgg_parallel_exec_tests.txt](https://github.com/tensorflow/tensorflow/files/1631436/vgg_parallel_exec_tests.txt)

I've also updated the information at the top. Working on a script I can share to reproduce.

**Comment by joeyearsley** [comment]
Additional info:
- Ran with `tf.data` api, NCCL replicated (tried all parameter servers) and staging areas on both the CPU and GPU.
- The model is a fully convolutional network with skip connections, like resnet.

**Comment by joeyearsley** [comment]
Would all of the *.dot files help? - Nearly finished code I can share (Other things to run in parallel.)

**Comment by jlebar** [comment]
The dot files wouldn't help a lot, but if you can dump a SessionModule proto (--xla_dump_computations_to=/tmp/foo, possibly passing the flag via the TF_XLA_FLAGS envvar), that is probably enough for us to reproduce.  You'd want to dump this from TF built at or near to HEAD, as that format changes, and our tools can't necessarily go back.

**Comment by joeyearsley** [comment]
[xlacomps.zip](https://github.com/tensorflow/tensorflow/files/1634595/xlacomps.zip)

Created with TF1.5 RC, please let me know if it needs to be any newer

**Comment by jlebar** [comment]
Just tried it, I don't think it's new enough, sadly.

**Comment by jlebar** [comment]
CUDA 9.1.85 is, I am told, the first public version to contain the fix to the ptxas problems we've been having.

Unfortunately I cannot vouch for CUDA 9.1 at all.

**Comment by jlebar** [comment]
Sorry, nvidia gave me incorrect information.  9.1.85 is known-bad, and there does not appear to be a fix publicly available at the moment.

**Comment by joeyearsley** [comment]
Thanks, I seemingly still had the problem with CUDA 8 and cudnn 6 also. 

Currently building a master container with 9/7 to get those computations and then will build an 8/6 one to follow it up (just to ensure it is/isn't ptxas).

**Comment by joeyearsley** [comment]
[xlacomputations.zip](https://github.com/tensorflow/tensorflow/files/1643208/xlacomputations.zip)

Here are the computations ran with TF Master, CUDA 8 and Cudnn 6, same proplem with the loss not going down.

**Comment by jlebar** [comment]
These SessionModules work, thanks.

It looks like I'm going to need to build some tools in order to figure out what's going wrong here.  Stay tuned, I guess...

**Comment by joeyearsley** [comment]
Thanks for keeping me in the loop!

**Comment by jlebar** [comment]
Spent a while debugging this today.  I've found a number of XLA computations which generate different values than we get with the reference backend, but this isn't necessarily a sign of a bug.  In particular, cudnn convolution isn't necessarily deterministic or even particularly accurate, depending on the algorithm we choose.

In theory XLA and TF use pretty much the same method for choosing cudnn algorithms, so I'd be surprised if this was the source of your problem.  But it's making it hard for me to tell what, if anything, is going wrong here...

I'm going to try to spend another day poking around at the XLA level to see if I can see what's going wrong here, but we may need that e2e testcase after all.

cc @hawkinsp, who may have ideas himself.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by jlebar** [comment]
I haven't been able to find anything suspicious in the SessionModules.  :(

That's not to say that something isn't broken -- I think something probably is.  But just that whatever is broken isn't obvious when I shove random data through these XLA parts.

A TF reproducer would very likely help.

**Comment by reedwm** [comment]
Marking as "awaiting response", since it appears a TF reproducer is needed to debug this.

As for the issue "XLA doesn't respect visible devices (neither CUDA Visible devices or tf config options)", @joeyearsley, can you file a separate issue?

**Comment by joeyearsley** [comment]
Hi @reedwm, I'll try and get the issue up once I've confirmed it on the latest release.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jlebar** [comment]
> CUDA 9.1.85 is, I am told, the first public version to contain the fix to the ptxas problems we've been having.

CUDA 9.1.85.2 has just been released, and it fixes this problem.

**Comment by joeyearsley** [comment]
@rapatel0 is looking into this on our end now, he'll get back hopefully by the end of the day (UK/GMT)

**Comment by rapatel0** [comment]
Tested with 9.1.85 and its broken but I didn't see that the patches were not applied in nvidia's default container. Will patch and get back to you

**Comment by rapatel0** [comment]
I added the cuda patches to my existing docker container and ran the test with the XLA streamexecutor running on CPU+GPU, GPU only, and a model without XLA as a baseline. We set the global seed in all cases and all image sizes and batch sizes are the same.
 
- After 70,000 iterations the noXLA model reaches a loss of 0.19  and is still trending downward.
- After 120,000 iterations both XLA models oscillate between losses of 0.3-0.4 and continue to oscillate until a limit of 200,000 iterations. 

I still need to try a test with patches applied before compiling tensorflow. Not sure if this will make a difference. It seems the convergence issue is still present.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jlebar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jlebar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by TylerBalsam** [comment]
Still getting bad shape problems on Cuda 9.1 where the filter sizes/strides/shapes are garbage w/ ptxas 9.1.121, Cudnn 7.1.2, Tensorflow built from source running on V100s:

Couldn't allocate space for input/filter/output of convolution %custom-call.24 = (f32[20,64,80,64]{3,2,1,0}, u8[0]{0}) custom-call(f32[20,2078,3238,64]{3,2,1,0} %pad.3, f32[125,157,64,64]{3,2,1,0} %multiply.62, s64[] %constant.161), window={size=125x157 stride=31x39}, dim_labels=b01f_01io->b01f, custom_call_target="__cudnn$convForward". Falling back to default algorithm.

Segmentation fault (core dumped)

Do you know if ptxas still having issues or is this a stable version and the problem is somewhere else in the XLA/JIT pipeline?

**Comment by jlebar** [comment]
@TylerBalsam , this error seems quite different from the other etiologies discussed in this bug.

"Couldn't allocate space" error is coming from cudnn_algorithm_picker.  It's trying to choose the fastest cudnn algorithm to use for the given convolution.  Which is not so surprising -- the result of the conv has size 20*2078*3238*64 * 4 bytes = 34gb, which is not going to fit on a V100.

I'm not sure what the segfault is about.  We shouldn't segfault on GPU OOM, but it's entirely possible the CPU is OOM'ing somewhere or something.  Not really enough info here to say.

**Comment by jlebar** [comment]
I'm not aware of any bugs in ptxas 9.1.121 that affect TF or XLA.

**Comment by jlebar** [comment]
I'm going to close this bug due to inactivity on the initially reported problem, but please feel free to reopen if that persists, and feel free to cc me on other XLA issues.
